pkgs/test_api/lib/hooks.dart

```diff
@@ -0,0 +1,75 @@
+// Copyright (c) 2021, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:stack_trace/stack_trace.dart';
+
+import 'src/backend/closed_exception.dart';
+import 'src/backend/invoker.dart';
+import 'src/backend/stack_trace_formatter.dart';
+
+class TestHandle {
+  /// Returns handle for the currently running test.
+  ///
+  /// This must be called from within the zone that the test is running in. If
+  /// the current zone is no a test's zone throws [OutsideTestException].
+  static TestHandle get current {
+    final invoker = Invoker.current;
+    if (invoker == null) throw OutsideTestException();
+    return TestHandle._(invoker, StackTraceFormatter.current);
+  }
+
+  final Invoker _invoker;
+  final StackTraceFormatter? _stackTraceFormatter;
+  TestHandle._(this._invoker, this._stackTraceFormatter);
+
+  String get name => _invoker.liveTest.test.name;
+
+  /// Whether this test has already completed successfully.
+  ///
+  /// If a callback originating from a test case is invoked after the test has
+  /// already passed it may be an indication of a test that fails to wait for
+  /// all work to be finished, or of an asynchronous callback that is called
+  /// more times or later than expected.
+  bool get shouldBeDone => _invoker.liveTest.state.shouldBeDone;
+
+  /// Marks this test as skipped.
+  ///
+  /// A skipped test may still fail if any exception is thrown, including
+  /// uncaught asynchronous errors.
+  void markSkipped(String message) {
+    if (_invoker.closed) throw ClosedException();
+    _invoker.skip(message);
+  }
+
+  /// Indicates that this test should to be considered done until [future]
+  /// completes.
+  ///
+  /// The test may time out before [future] completes.
+  Future<T> mustWaitFor<T>(Future<T> future) {
```

Comment:
I suspect this may have issues if `future` comes from a different error zone than the test. We can't detect this mistake either. https://github.com/dart-lang/sdk/issues/39825#issuecomment-799862584

I still think it's worth having this API over the increment/decrement calls because with some refactoring we could hopefully make some problems easier to debug (make it possible to know _which_ futures haven't completed yet and investigate them in observatory, as opposed to only seeing that a counter hasn't reached zero).

I think probably anyone who needs to understand this API is going to need to understand some of the details of error zones anyway, so this might not be a big problem. I'm curious if @jakemac53 or @grouma is concerned though.

Reply:
> (make it possible to know which futures haven't completed yet and investigate them in observatory, as opposed to only seeing that a counter hasn't reached zero)

Anything that helps towards this goal is great in my opinion. This has wasted so many hours of my life....

Reply:
OK, looks like this isn't a new issue.

https://github.com/dart-lang/test/issues/1468

It turns out my design here actually fixes that issue, which I _think_ is a good thing. I'm a tiny bit concerned about unintended consequences, but my hunch is it will be an improvement.

Reply:
In case it isn't clear, here is a small repro. This test fails currently, and with this PR it will pass.

```dart
import 'dart:async';

import 'package:test/test.dart';

void main() {
  test('runs an async outside of the test zone', () async {
    var x = expectAsync0(() {});
    Zone.root.run(x);
  });
}
```

If anyone has concerns about this let me know.

Reply:
There is another case that this impacts, but doesn't fix.

If a test _fails_ in an async callback that is running in the "wrong" zone, we don't still don't show the failure nicely. Consider the following:

```dart
import 'dart:async';

import 'package:test/test.dart';

void main() {
  test('fails an async outside of the test zone', () async {
    var x = expectAsync0(() {
      expect(true, false);
    });
    Zone.root.run(x);
  });
}
```

Before this would fail like:

```
00:00 +0 -1: fails an async outside of the test zone [E]
  Instance of 'OutsideTestException'
  package:test_api/hooks.dart 18:26                              TestHandle.current
  package:test_api/src/expect/expect.dart 70:27                  _expect
  package:test_api/src/expect/expect.dart 46:3                   expect
  test/repro_test.dart 8:7                                       main.<fn>.<fn>
  dart:core                                                      Function.apply
  package:test_api/src/expect/expect_async.dart 196:23           _ExpectedFunction._run
  package:test_api/src/expect/expect_async.dart 179:7            _ExpectedFunction.max6
  package:test_api/src/expect/expect_async.dart 145:15           _ExpectedFunction.max0
  dart:async                                                     _RootZone.run
  test/repro_test.dart 10:15                                     main.<fn>
  test/repro_test.dart 6:51                                      main.<fn>
  package:test_api/src/backend/declarer.dart 200:19              Declarer.test.<fn>.<fn>
  ===== asynchronous gap ===========================
  dart:async                                                     _asyncErrorWrapperHelper
  package:test_api/src/backend/invoker.dart                      Invoker.waitForOutstandingCallbacks.<fn>
  dart:async                                                     runZoned
  package:test_api/src/backend/invoker.dart 228:5                Invoker.waitForOutstandingCallbacks
  package:test_api/src/backend/invoker.dart 383:17               Invoker._onRun.<fn>.<fn>.<fn>
  ===== asynchronous gap ===========================
  dart:async                                                     _asyncThenWrapperHelper
  package:test_api/src/backend/invoker.dart                      Invoker._onRun.<fn>.<fn>.<fn>
  dart:async                                                     runZoned
  package:test_api/src/backend/invoker.dart 370:9                Invoker._onRun.<fn>.<fn>
  package:test_api/src/backend/invoker.dart 415:15               Invoker._guardIfGuarded
  package:test_api/src/backend/invoker.dart 369:7                Invoker._onRun.<fn>
  package:stack_trace                                            Chain.capture
  package:test_api/src/backend/invoker.dart 368:11               Invoker._onRun
  package:test_api/src/backend/live_test_controller.dart 153:11  LiveTestController.run
  package:test_api/src/remote_listener.dart 255:16               RemoteListener._runLiveTest.<fn>
  dart:async                                                     runZoned
  package:test_api/src/remote_listener.dart 254:5                RemoteListener._runLiveTest
  package:test_api/src/remote_listener.dart 207:7                RemoteListener._serializeTest.<fn>
 ```
 
 Whereas now it fails like
 
 ```
 00:01 +0 -1: fails an async outside of the test zone [E]
  Bad state: Can't add or remove outstanding callbacks outside of a test body.
  package:test_api/src/backend/invoker.dart 119:5         Invoker._outstandingCallbacks
  package:test_api/src/backend/invoker.dart 215:5         Invoker.removeOutstandingCallback
  package:test_api/src/frontend/expect_async.dart 208:15  _ExpectedFunction._afterRun
  package:test_api/src/frontend/expect_async.dart 195:7   _ExpectedFunction._run
  package:test_api/src/frontend/expect_async.dart 176:7   _ExpectedFunction.max6
  package:test_api/src/frontend/expect_async.dart 142:15  _ExpectedFunction.max0
  dart:async                                              _RootZone.run
  test/repro_test.dart 10:15                              main.<fn>
  test/repro_test.dart 6:51                               main.<fn>
  package:test_api/src/backend/declarer.dart 200:19       Declarer.test.<fn>.<fn>
  ===== asynchronous gap ===========================
  package:test_api/src/backend/declarer.dart 198:7        Declarer.test.<fn>
  ===== asynchronous gap ===========================
  package:test_api/src/backend/invoker.dart 231:7         Invoker.waitForOutstandingCallbacks.<fn>
 ```
 
 This feels like a wash to me so I don't think it should be blocking. I'll see if I can solve this case too though.

Reply:
> I suspect this may have issues if `future` comes from a different error zone than the test. We can't detect this mistake either.

I can't actually find any situation where this causes a problem. I think we are OK on that part.